fecgov/fec-cms
# Error handling for about/careers (USAJobs API)

## Summary (required)

- Resolves #6260
- Expand the error handling for ConnectionErrors not caught by the current check for 200 http response,
- Expand the error handling to catch errors that have a 200 http response, but can still break the page.
    - When the request returns html or json data that is not expected jobs dict which which previously caused a `500 Internal server error`
- Catch connection errors if USAJobs API is not responding or down for maintenance 
- Write the error class name to the page to give devs an indication of what type of error is happening

### Required reviewers

One python developer

## Impacted areas of the application

`/about/careers/`
`open_jobs.py - get_jobs()` templatetag


## How to test

- Checkout and run branch
- Export the following  env vars from CF in your shell :
    - USAJOBS_API_KEY
    - USAJOBS_WHOMAYAPPLY
    - USAJOBS_AGENCY_CODE (since we currently have no open jobs, you can use `PU00` for `Peace Corps` if you want to confirm job data renders to the page)
 - view the page at `/about/careers/`
 - run `pytest` from /fec-cms and /fec-cms/fec directories
 
 Below are a few ways to trigger various error scenarios if you are interested in testing them:
 - Try changing the JOB_URL in `open_jobs.py` to a non existent website. Should get (ConnectionError) at the end of the error message in the page.
 - Try adding `querystring["page"] = 2` to the querystring parameters, should see: "There are currently no open positions..." message (since it shows a max of 25 jobs per page, it will return no jobs)
  - Comment out `"host": "data.usajobs.gov"` (@ line 36) and then change the JOB_URL to `"https://dummyjson.com/products"`, should get a  (KeyError)
